### ISSUE TITLE
Decouple element variable name in HTML template definitions

### ISSUE THREAD

**Issue by treshugart** [tmpfs/trucks#9]
I was reading through the README and saw in the template definitions you use `${elem.title}` which maps directly to the `elem` argument to `render()`. I was wondering if you thought it'd be a good idea to remove the need to prefix variables with `elem.` and instead specify only `title`. The compiler could then check the name of the variable and auto-prefix them.

**Comment by tmpfs** [comment]
I like the conciseness of it of course but from an implementation point of view I don't think it's trivial as any javascript expression can be in the template literal I would need to find out the properties defined by the component javascript but as there are so many different ways of defining skate components (object literal, ES6 class etc) I'm potentially dealing with quite a few different AST structures to determine the valid list of properties (that are attributes).

But if say we enforced that the author declared an example component usage as per the [documentation roadmap](https://github.com/tmpfs/trucks#documentation) then I could read the list of attributes from the example component declaration. I think providing example usage would be considered best practice.

As an aside, template literal support is not enabled by default; I've been thinking the more common use case would be to use template literals, do you think it should be enabled by default?

**Comment by treshugart** [comment]
Ah ok, good point. I wasn't sure if you were allowing that or not. I like the template literal idea because the parsing is simple and it's an ES spec. Having them embedded in the HTML raises questions. For example how would you hide / show elements?

For example:

```js
<template id="x-blog-post">
  <div class="post">
    <h3>${elem.title}</h3>
    ${elem.content ? '<p>${elem.content}</p>' : ''}
  </div>
</template>
```

That would be a custom parse job, most likely. If you're just passing text content into template literals, then you might be limited to that unless you introduce custom semantics.

FWIW, this is why I generally opt to use JSX if I want to describe my templates in an HTML-like syntax even though it requires a build-step.

**Comment by tmpfs** [comment]
Thanks, that's the sort of thing I thought I'd be missing, I think passing through a JSX transformer is definitely an option.

The other things that I considered are a tagged template literal:

```
html`${elem.content} ? '...' :'...'}`
```

Which could get executed at compile time.

As well as possibly allowing the `<template>` element to be passed through an arbitrary template library (JSX, handlebars, jade etc).

Which do you prefer? I think the tagged template literal is cool again because it's ES but might not be as flexible as allowing a template preprocessor.

**Comment by tmpfs** [comment]
Regarding your original point we could probably modify the scope of the function call in the template main entry point so that rather than `${elem.title}` we could do `${this.title}` to refer to component properties. Seems a little more elegant to me...

**Comment by treshugart** [comment]
+1 for `this`

I'm kinda torn. Tagged template literals are close to specs but JSX [has a spec](https://facebook.github.io/jsx/), too, it's just not TC39 official. It was designed for HTML whereas template literals are lower-level, and designed abstractly for strings of all types.

I don't have a strong opinion, really. I think either way you go you have a good philosophy of staying close to specs and keeping things simple. I look forward to trying it soon :)

**Comment by treshugart** [comment]
Feel free to close this if you like.

**Comment by tmpfs** [comment]
Implemented referencing properties via `this` in f7761ff21937a50ea933805ce7fcbd3210975559.

Closing as implemented, I'm going to keep thinking about the ways of addressing conditional HTML elements in the templates.
